File: src/main/java/net/database/RankingAPI.java
package net.database;

import org.jooq.DSLContext;
import org.jooq.Record;
import org.jooq.Result;
import world.ranking.PlayerRanking;

import java.util.*;

import static database.jooq.Tables.CHARACTERS;
import static database.jooq.Tables.MOBKILLS;

public class RankingAPI {

    public static Map<String, PlayerRanking> getNonGMCharacters() {
        DSLContext con = DatabaseCore.getConnection();
        Map<String, PlayerRanking> rankings = new LinkedHashMap<>();

        Result<Record> res = con.select().from(CHARACTERS)
                .leftJoin(MOBKILLS)
                .onKey()
                .where(CHARACTERS.GM_LEVEL.eq(0))
                .fetch();

        res.forEach(rec -> {
            String name = rec.getValue(CHARACTERS.NAME);
            PlayerRanking ranking = rankings.get(name);
            if (ranking == null) {
                ranking = new PlayerRanking(
                        rec.getValue(CHARACTERS.LEVEL),
                        rec.getValue(CHARACTERS.JOB),
                        rec.getValue(CHARACTERS.KILL_COUNT),
                        name,
                        rec.getValue(CHARACTERS.HARDCORE) == 1,
                        rec.getValue(CHARACTERS.MAP) == 666);
                rankings.put(name, ranking);
            }
            if (rec.getValue(MOBKILLS.MID) != null)
                ranking.getMobKills().put(rec.getValue(MOBKILLS.MID), rec.getValue(MOBKILLS.COUNT));
        });

        return rankings;
    }
}

File: src/main/java/net/database/RankingAPI.kt
package net.database

import database.jooq.Tables.CHARACTERS
import database.jooq.Tables.MOBKILLS
import net.database.DatabaseCore.connection
import world.ranking.PlayerRanking

object RankingAPI {

    val nonGMCharacters: Map<String, PlayerRanking>
        get() {
            val rankings: MutableMap<String, PlayerRanking> = LinkedHashMap()
            val res = connection.select().from(CHARACTERS)
                    .leftJoin(MOBKILLS)
                    .onKey()
                    .where(CHARACTERS.GM_LEVEL.eq(0))
                    .fetch()

            res.forEach {
                val name = it.getValue(CHARACTERS.NAME)
                var ranking = rankings[name]
                if (ranking == null) {
                    ranking = PlayerRanking(
                            it.getValue(CHARACTERS.LEVEL),
                            it.getValue(CHARACTERS.JOB),
                            it.getValue(CHARACTERS.KILL_COUNT),
                            name,
                            it.getValue(CHARACTERS.HARDCORE) == 1.toByte(),
                            it.getValue(CHARACTERS.MAP) == 666)
                    rankings[name] = ranking
                }

                if (it.getValue(MOBKILLS.MID) != null)
                    ranking.mobKills[it.getValue(MOBKILLS.MID)] = it.getValue(MOBKILLS.COUNT)
            }
            return rankings
        }
}